impl/src/main/java/com/sun/faces/lifecycle/ClientWindowImpl.java

```diff
@@ -64,9 +65,9 @@ private String calculateClientWindow(FacesContext context) {
                 counter = Integer.valueOf(0);
             }
             char sep = UINamingContainer.getSeparatorChar(context);
-            id = extContext.getSessionId(true) + sep + +counter;
+            id = tokenGenerator.getNextToken() + sep + ++counter;
```

Comment:
Not super familiar with the client window, if we are doing a token + counter stored with the session shouldn't the token be stored with the session as well so that we have a consistent token per session?

Do we even need a token? or can it just increment like was done for conversation id? The client window id is only valid with the users session anyway right?

Reply:
I'm not expert on this either, but the value can't be stored in the session, as there can be multiple client windows at any given time. The value is retrieved from the request in the decode() method, with this method only being called when the ID can't be retrieved from the request.